src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -28,17 +28,17 @@ import scala.util.control.{NoStackTrace, NonFatal}
  *
  * TODO: move out of `analysis` package?
  */
-abstract class BackendUtils extends PerRunLazy {
+abstract class BackendUtils extends PerRunInit {
   val postProcessor: PostProcessor
 
-  import postProcessor.{bTypes, bTypesFromClassfile, callGraph, frontendAccess}
+  import postProcessor.{bTypes, bTypesFromClassfile, callGraph}
   import bTypes._
   import callGraph.ClosureInstantiation
   import coreBTypes._
   import frontendAccess.compilerSettings
 
   // unused objects created by these constructors are eliminated by pushPop
-  private[this] lazy val sideEffectFreeConstructors: LazyVar[Set[(String, String)]] = perRunLazy {
+  private[this] lazy val sideEffectFreeConstructors: LazyVar[Set[(String, String)]] = perRunLazy(this) {
```

Comment:
There's probably too many implicit invariants, but here they are:
  - `initialize` is called in each compiler run at the beginning when the GenBCode phase starts running. There's no multithreading at this point.
  - I kept the values here as lazy vals so that only those actually accessed are ever added to the `PerRunInit.this.inits` buffer (and therefore re-initializes in the next run).
  - I *think* there can't be a deadlock here. Initializing a lazy val here will create the `LazyVar` and add it to the `this.inits` buffer, and this happens synchronized on `this`.
  - Completing a `LazyVar` synchronizes on the `frontendLock`.